File: app/src/main/java/com/example/sunkai/heritage/MainActivity.java
package com.example.sunkai.heritage;

import android.app.ActionBar;
import android.content.DialogInterface;
import android.graphics.Color;
import android.os.Bundle;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.view.KeyEvent;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.ImageView;
import android.widget.TextView;

/**
 * 此类用于处理登陆
 */

public class MainActivity extends AppCompatActivity implements OnClickListener{
    /**
     * 用于展示消息的Fragment
     */
    private MainFragment mainFragment;

    /**
     * 用于展示联系人的Fragment
     */
    private folkFragment folkFragment;

    /**
     * 用于展示动态的Fragment
     */
    private findFragment findFragment;

    /**
     * 用于展示设置的Fragment
     */
    private personFragment personFragment;

    /**
     * 消息界面布局
     */
    private View mainLayout;

    /**
     * 联系人界面布局
     */
    private View folkLayout;

    /**
     * 动态界面布局
     */
    private View findLayout;

    /**
     * 设置界面布局
     */
    private View personLayout;

    /**
     * 在Tab布局上显示消息图标的控件
     */
    private ImageView mainImage;

    /**
     * 在Tab布局上显示联系人图标的控件
     */
    private ImageView folkImage;

    /**
     * 在Tab布局上显示动态图标的控件
     */
    private ImageView findImage;

    /**
     * 在Tab布局上显示设置图标的控件
     */
    private ImageView personImage;

    /**
     * 在Tab布局上显示消息标题的控件
     */
    private TextView mainText;

    /**
     * 在Tab布局上显示联系人标题的控件
     */
    private TextView folkText;

    /**
     * 在Tab布局上显示动态标题的控件
     */
    private TextView findText;

    /**
     * 在Tab布局上显示设置标题的控件
     */
    private TextView personText;

    /**
     * 用于对Fragment进行管理
     */
    private android.support.v4.app.FragmentManager fragmentManager;


    AlertDialog ad;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        initViews();
        fragmentManager=getSupportFragmentManager();
        setTabSelection(0);
//0        ActionBar actionBarr = getActionBar();
//        actionBar.hide();
    }

    private void initViews() {
        mainLayout = findViewById(R.id.main_layout);
        folkLayout = findViewById(R.id.folk_layout);
        findLayout = findViewById(R.id.find_layout);
        personLayout = findViewById(R.id.person_layout);
        mainImage = (ImageView) findViewById(R.id.main_image);
        folkImage = (ImageView) findViewById(R.id.folk_image);
        findImage = (ImageView) findViewById(R.id.find_image);
        personImage = (ImageView) findViewById(R.id.person_image);
        mainText = (TextView) findViewById(R.id.main_text);
        folkText = (TextView) findViewById(R.id.folk_text);
        findText = (TextView) findViewById(R.id.find_text);
        personText = (TextView) findViewById(R.id.person_text);
        mainLayout.setOnClickListener(this);
        folkLayout.setOnClickListener(this);
        findLayout.setOnClickListener(this);
        personLayout.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.main_layout:
                // 当点击了消息tab时，选中第1个tab
                setTabSelection(0);
                break;
            case R.id.folk_layout:
                // 当点击了联系人tab时，选中第2个tab
                setTabSelection(1);
                break;
            case R.id.find_layout:
                // 当点击了动态tab时，选中第3个tab
                setTabSelection(2);
                break;
            case R.id.person_layout:
                // 当点击了设置tab时，选中第4个tab
                setTabSelection(3);
                break;
            default:
                break;
        }
    }

    private void setTabSelection(int index) {
        clearSelection();

        FragmentTransaction transaction = fragmentManager.beginTransaction();

        hideFragments(transaction);

        switch (index) {
            case 0:
                mainImage.setImageResource(R.drawable.ic_home_brown_500_24dp);
                mainText.setTextColor(Color.BLACK);
                if (null == mainFragment) {
                    mainFragment = new MainFragment();
                    transaction.add(R.id.content, mainFragment);
                } else {
                    transaction.show(mainFragment);
                }
                break;
            case 1:
                folkImage.setImageResource(R.drawable.ic_people_brown_500_24dp);
                folkText.setTextColor(Color.BLACK);
                if (null == folkFragment) {
                    folkFragment = new folkFragment();
                    transaction.add(R.id.content, folkFragment);
                } else {
                    transaction.show(folkFragment);
                }
                break;
            case 2:
                findImage.setImageResource(R.drawable.ic_search_brown_500_24dp);
                findText.setTextColor(Color.BLACK);
                if (null == findFragment) {
                    findFragment = new findFragment();
                    transaction.add(R.id.content, findFragment);
                } else {
                    transaction.show(findFragment);
                }
                break;
            case 3:
            default:
                personImage.setImageResource(R.drawable.ic_assignment_ind_brown_500_24dp);
                personText.setTextColor(Color.BLACK);
                if (null == personFragment) {
                    personFragment = new personFragment();
                    transaction.add(R.id.content, personFragment);
                } else {
                    transaction.show(personFragment);
                }
                break;
        }
        transaction.commit();
    }

    private void clearSelection(){
        mainImage.setImageResource(R.drawable.ic_home_grey_500_24dp);
        mainText.setTextColor(Color.parseColor("#82858b"));
        folkImage.setImageResource(R.drawable.ic_people_grey_500_24dp);
        folkText.setTextColor(Color.parseColor("#82858b"));
        findImage.setImageResource(R.drawable.ic_search_grey_500_24dp);
        findText.setTextColor(Color.parseColor("#82858b"));
        personImage.setImageResource(R.drawable.ic_assignment_ind_grey_500_24dp);
        personText.setTextColor(Color.parseColor("#82858b"));
    }

    private void hideFragments(FragmentTransaction transaction){
        if(null!=mainFragment){
            transaction.hide(mainFragment);
        }
        if(null!=folkFragment){
            transaction.hide(folkFragment);
        }
        if(null!=findFragment){
            transaction.hide(findFragment);
        }
        if(null!=personFragment){
            transaction.hide(personFragment);
        }
    }


    //重写onKeyDown方法，监听返回键
    public boolean onKeyDown(int keyCode, KeyEvent event)
    {
        if(keyCode==KeyEvent.KEYCODE_BACK) {

            AlertDialog.Builder builder = new AlertDialog.Builder(MainActivity.this).setTitle("退出?").setPositiveButton("退出", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        ad.dismiss();
                        finish();
                    }
                }).setNegativeButton("取消", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        ad.dismiss();
                    }
                });
                ad = builder.create();
                ad.show();
                return super.onKeyDown(keyCode, event);
        }
        return super.onKeyDown(keyCode, event);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        WelcomeActivity.myHelper.close();
    }
}

File: app/src/main/java/com/example/sunkai/heritage/Activity/MainActivity.kt
package com.example.sunkai.heritage.Activity

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.content.res.ColorStateList
import android.os.Build
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import android.view.KeyEvent
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager
import com.example.sunkai.heritage.Activity.BaseActivity.BaseGlideActivity
import com.example.sunkai.heritage.Dialog.NormalWarningDialog
import com.example.sunkai.heritage.Fragment.BaseFragment.BaseLazyLoadFragment
import com.example.sunkai.heritage.Fragment.FindFragment
import com.example.sunkai.heritage.Fragment.FolkFragment
import com.example.sunkai.heritage.Fragment.MainFragment
import com.example.sunkai.heritage.Fragment.PersonFragment
import com.example.sunkai.heritage.R
import com.example.sunkai.heritage.Service.PushService
import com.example.sunkai.heritage.tools.getDarkThemeColor
import com.example.sunkai.heritage.tools.getThemeColor
import com.example.sunkai.heritage.value.PUSH_SWITCH
import com.example.sunkai.heritage.value.SETTING
import com.example.sunkai.heritage.value.TYPE_FOLK_HERITAGE
import com.google.android.material.bottomnavigation.BottomNavigationView
import kotlinx.android.synthetic.main.activity_main.*
import java.lang.ref.WeakReference

/**
 * 此类用于处理登陆
 */

class MainActivity : BaseGlideActivity() {

    //让其他Activity可以访问MainActivity的方法，提供了一个弱引用
    companion object {
        var activityRef: WeakReference<MainActivity>? = null
        var bottomNavigationRef: WeakReference<BottomNavigationView>? = null
        fun GetViewpagerSelectPosition(): Int {
            val activity = activityRef?.get() ?: return 0
            //防止找不到view的容错，加了个？ 但看了转换的Java代码，似乎没用？
            return activity.bottomNavigationButton.selectedItemId
        }
    }

    private val viewList: ArrayList<Fragment> = ArrayList()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        activityRef = WeakReference(this)
        setContentView(R.layout.activity_main)
        initViews()
        //如果用户打开了推送开关，则开启推送服务
        val sharePrefrence = getSharedPreferences(SETTING, Context.MODE_PRIVATE)
        if (sharePrefrence.getBoolean(PUSH_SWITCH, false)) {
            startPushService()
        }
        setIgnoreToolbar(true)
    }


    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        //当window建立起来之后，初始化里面的view
        bottomNavigationRef = WeakReference(bottomNavigationButton)
    }

    private var mBoundService: PushService? = null
    private var mShouldBind = false
    private val mConnection = object : ServiceConnection {
        override fun onServiceDisconnected(p0: ComponentName?) {
            mBoundService = null
        }

        override fun onServiceConnected(p0: ComponentName?, p1: IBinder?) {
            mBoundService = (p1 as PushService.LocalBinder).service
        }
    }

    fun startPushService() {
        if (bindService(Intent(this, PushService::class.java), mConnection, Context.BIND_AUTO_CREATE)) {
            mShouldBind = true
        }
    }

    override fun setNeedChangeThemeColorWidget() {
        changeThemeWidge.add(R.id.activityMainViewpager)
    }

    private fun initViews() {
        val themeColor = getThemeColor()
        val midGrey = ContextCompat.getColor(this@MainActivity, R.color.midGrey)
        val colors = arrayOf(themeColor, midGrey).toIntArray()
        val states = arrayOf(arrayOf(android.R.attr.state_checked).toIntArray(), arrayOf(-android.R.attr.state_checked).toIntArray())
        val colorStateList = ColorStateList(states, colors)
        bottomNavigationButton.itemTextColor = colorStateList
        bottomNavigationButton.itemIconTintList = colorStateList
        bottomNavigationButton.setOnNavigationItemSelectedListener(onNavigationItemSelectedListener)
        viewList.add(MainFragment())
        viewList.add(FolkFragment())
        viewList.add(FindFragment())
        viewList.add(PersonFragment())
        val adapter = adapter(viewList, supportFragmentManager)
        activityMainViewpager.adapter = adapter
        activityMainViewpager.addOnPageChangeListener(onPageChangeListener)
    }

    //重写onKeyDown方法，监听返回键
    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            showExitDialog()
            return true
        }
        return super.onKeyDown(keyCode, event)
    }

    private fun showExitDialog() {
        NormalWarningDialog().setTitle("退出?")
                .setOnSubmitClickListener(object : NormalWarningDialog.onSubmitClickListener {
                    override fun onSubmit(view: View, dialog: NormalWarningDialog) {
                        dialog.dismiss()
                        finish()
                    }
                })
                .setSubmitText("退出")
                .show(supportFragmentManager, "exitDialog")
    }

    override fun onDestroy() {
        doUnbindService()
        stopService(Intent(this, PushService::class.java))
        super.onDestroy()
    }

    fun doUnbindService() {
        if (mShouldBind) {
            unbindService(mConnection)
            mShouldBind = false
        }
    }

    private class adapter(val viewList: ArrayList<Fragment>, manager: FragmentManager) : FragmentPagerAdapter(manager) {
        override fun getItem(position: Int): Fragment {
            return viewList[position]
        }

        override fun getCount(): Int {
            return viewList.size
        }

        override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {

        }
    }

    private val onPageChangeListener = object : ViewPager.OnPageChangeListener {
        override fun onPageScrollStateChanged(state: Int) {
        }

        override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
        }

        override fun onPageSelected(position: Int) {
            val themeColor = getThemeColor()
            val midGrey = ContextCompat.getColor(this@MainActivity, R.color.midGrey)
            val colors = arrayOf(themeColor, midGrey).toIntArray()
            val states = arrayOf(arrayOf(android.R.attr.state_checked).toIntArray(), arrayOf(-android.R.attr.state_checked).toIntArray())
            val colorStateList = ColorStateList(states, colors)
            bottomNavigationButton.itemTextColor = colorStateList
            bottomNavigationButton.itemIconTintList = colorStateList
            if (Build.VERSION.SDK_INT >= 21) {
                window.navigationBarColor = themeColor
                window.statusBarColor = when (position) {
                    1 -> {
                        //当从别的页面进入民间页的时候，转换状态栏和底栏的颜色
                        val color = (viewList[position] as FolkFragment).getStatusBarShouldChangeColor()
                        window.navigationBarColor = color
                        val folkColors = arrayOf(color, midGrey).toIntArray()
                        val folkStates = arrayOf(arrayOf(android.R.attr.state_checked).toIntArray(), arrayOf(-android.R.attr.state_checked).toIntArray())
                        val folkColorStateList = ColorStateList(folkStates, folkColors)
                        bottomNavigationButton.itemTextColor = folkColorStateList
                        bottomNavigationButton.itemIconTintList = folkColorStateList
                        color
                    }
                    else -> {
                        getDarkThemeColor()
                    }
                }
                val adapter = activityMainViewpager.adapter
                if (adapter is adapter) {
                    val fragment = adapter.getItem(position)
                    if (fragment is BaseLazyLoadFragment) {
                        fragment.lazyLoad()
                    }
                }
                bottomNavigationButton.selectedItemId = when (position) {
                    0 -> R.id.main_layout
                    1 -> R.id.folk_layout
                    2 -> R.id.find_layout
                    3 -> R.id.person_layout
                    else -> R.id.main_layout
                }
                when (position) {
                    0 -> {
                        window.setBackgroundDrawableResource(R.color.white)
                    }
                    else -> {
                        window.setBackgroundDrawable(null)
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                            window.exitTransition = null
                        }
                    }
                }
            }
        }

    }
    private val onNavigationItemSelectedListener = BottomNavigationView.OnNavigationItemSelectedListener { item ->
        activityMainViewpager.currentItem = when (item.itemId) {
            R.id.main_layout -> 0
            R.id.folk_layout -> 1
            R.id.find_layout -> 2
            R.id.person_layout -> 3
            else -> 0
        }
        true
    }
}
